internal/network/nioc_bandwidth_allocation_subresource.go

```diff
@@ -0,0 +1,106 @@
+/*
+ *  Copyright 2023 VMware, Inc.
+ *    SPDX-License-Identifier: MPL-2.0
+ */
+
+package network
+
+import (
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	validation_utils "github.com/vmware/terraform-provider-vcf/internal/validation"
+	"github.com/vmware/vcf-sdk-go/models"
+	"strings"
+)
+
+// NiocBandwidthAllocationSchema this helper function extracts the NiocBandwidthAllocation
+// Schema, so that it's made available for both Domain and Cluster creation.
+func NiocBandwidthAllocationSchema() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"type": {
+				Type:     schema.TypeString,
+				Required: true,
+				Description: "Host infrastructure traffic type. " +
+					"Example: management, faultTolerance, vmotion, " +
+					"virtualMachine, iSCSI, nfs, hbr, vsan, vdp etc.",
+				ValidateFunc: validation.NoZeroValues,
+			},
+			"limit": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Description: "The maximum allowed usage for a traffic class belonging to this resource pool per host " +
+					"physical NIC. The utilization of a traffic class will not exceed the specified limit even if " +
+					"there are available network resources. If this value is unset or set to -1 in an update " +
+					"operation, then there is no limit on the network resource usage (only bounded by available " +
+					"resource and shares). Units are in Mbits/sec",
+				ValidateFunc: validation.NoZeroValues,
+			},
+			"reservation": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Description: "Amount of bandwidth resource that is guaranteed available to the host infrastructure traffic " +
+					"class. If the utilization is less than the reservation, the extra bandwidth is used for other " +
+					"host infrastructure traffic class types. Unit is Mbits/sec",
+				ValidateFunc: validation.NoZeroValues,
+			},
+			"shares": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Description: "The number of shares allocated. Used to determine resource allocation in case of resource " +
+					"contention. This value is only set if level is set to custom. If level is not set to custom, " +
+					"this value is ignored. Therefore, only shares with custom values can be compared. " +
+					"There is no unit for this value. It is a relative measure based on the settings for other resource pools.",
+				ValidateFunc: validation.NoZeroValues,
+			},
+			"shares_level": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Description: "The allocation level. The level is a simplified view of shares. Levels map to a " +
+					"pre-determined set of numeric values for shares. If the shares value does not map to a " +
+					"predefined size, then the level is set as custom. One among: low, normal, high, custom",
+				ValidateFunc: validation.StringInSlice([]string{
+					"low", "normal", "high", "custom",
```

Comment:
Should this be moved to a var and then:


```go
ValidateFunc: validation.StringInSlice(networkShareLevelAllowedValues, false),
```

Reply:
as mentioned above, var can be changed, []string cannot be a constant...